# About this file

The file `hotel_bookings.csv` contains data on hotel bookings, including information on guest demographics, booking details, and reservation status.

Column Descriptions

- **hotel**: Indicates the type of hotel (resort or city). (Categorical)
- **is_canceled**: Specifies whether the booking was canceled or not (0=not canceled, 1=canceled). (Binary)
- **lead_time**: Represents the number of days between the booking date and the arrival date. (Numerical)
- **arrival_date_year**: Denotes the year of the arrival date. (Categorical)
- **arrival_date_month**: Indicates the month of the arrival date. (Categorical)
- **arrival_date_week_number**: Specifies the week number in which guests arrived at the hotel. (Numerical)
- **arrival_date_day_of_month**: Represents a specific day of arrival within a month. (Numerical)
- **stays_in_weekend_nights**: Indicates how many nights (Saturday or Sunday) guests stayed or booked to stay at a hotel during weekends. (Numerical)
- **stays_in_week_nights**: Represents how many weeknights (Monday to Friday) guests stayed or booked to stay at a hotel during weekdays. (Numerical)
- **adults**: Indicates the number of adults included in each booking. (Numerical)
- **children**: Indicates the number of children included in each booking. (Numerical)
- **babies**: Indicates the number of babies included in each booking. (Numerical)
- **meal**: Describes what type of meal was booked (Breakfast only, Half board, Full board, or Undefined/SC – no meal package). (Categorical)
- **country**: Denotes the country-of-origin for each guest who made a reservation. (Categorical)
- **market_segment**: Shows various market segments that individuals belong to when making reservations (e.g., Online Travel Agents, Offline Travel Agents, Corporate clients). (Categorical)
- **distribution_channel**: Specifies different channels through which bookings were made (e.g., online travel agencies, direct bookings with hotels/tour operators, corporate arrangements). (Categorical)
- **is_repeated_guest**: Indicates whether the guest is a repeated visitor (0=not repeated guest, 1=repeated guest). (Binary)
- **previous_cancellations**: Represents the number of times guests previously canceled their bookings. (Numerical)
- **previous_bookings_not_canceled**: Denotes the count of previous bookings made by guests that were not canceled. (Numerical)
- **reserved_room_type**: Identifies the type of room initially reserved. (Categorical)
- **assigned_room_type**: Identifies the type of room that was assigned to guests. (Categorical)
- **booking_changes**: Represents the number of changes made to the booking. (Numerical)
- **deposit_type**: Indicates the type of deposit made for the booking. (Categorical)
- **agent**: Represents the ID of the travel agency that made the booking. (Categorical)
- **company**: Represents the ID of the company that made the booking. (Categorical)
- **days_in_waiting_list**: Represents the number of days the booking was on the waiting list before being confirmed. (Numerical)
- **customer_type**: Indicates the type of customer (e.g., transient, contract, group, or other). (Categorical)
- **adr**: Represents the average daily rate (price per room) for the booking. (Numerical)
- **required_car_parking_spaces**: Indicates the number of car parking spaces required by the guest. (Numerical)
- **total_of_special_requests**: Represents the total number of special requests made by the guest (e.g., extra bed, room amenities). (Numerical)
- **reservation_status**: Indicates the status of the reservation (e.g., canceled, checked-in, no-show). (Categorical)
- **reservation_status_date**: Represents the date on which the reservation status was last updated. (Date)

# Import Data & Library

In [41]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.style as style
import matplotlib.ticker as mtick
from sklearn.model_selection import train_test_split, RepeatedStratifiedKFold, cross_val_score, RandomizedSearchCV
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.pipeline import Pipeline
from scipy.stats import chi2_contingency
from sklearn.feature_selection import f_classif
from sklearn.metrics import matthews_corrcoef
from sklearn.metrics import ConfusionMatrixDisplay
from sklearn.decomposition import PCA
%matplotlib inline


print('numpy version : ',np.__version__)
print('pandas version : ',pd.__version__)
print('seaborn version : ',sns.__version__)

numpy version :  1.24.3
pandas version :  1.5.3
seaborn version :  0.12.2


In [42]:
df = pd.read_csv('hotel_bookings_data.csv')
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2017          September   
1  Resort Hotel            0        737               2017          September   
2  Resort Hotel            0          7               2017          September   
3  Resort Hotel            0         13               2017          September   
4  Resort Hotel            0         14               2017          September   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_weekdays_nights  adults  ...  \
0                        0                         0       2  ...   
1                        0                         0       2  ...   
2                        0                         1       1  ...   
3                        0                         1       1  ...   
4                        0                         2       2  ...   

   booking_changes  deposit_type  agent company days_in_waiting_list  \
0                3    No Deposit    NaN     NaN                    0   
1                4    No Deposit    NaN     NaN                    0   
2                0    No Deposit    NaN     NaN                    0   
3                0    No Deposit  304.0     NaN                    0   
4                0    No Deposit  240.0     NaN                    0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0      Personal   0.0                            0                          0   
1      Personal   0.0                            0                          0   
2      Personal  75.0                            0                          0   
3      Personal  75.0                            0                          0   
4      Personal  98.0                            0                          1   

   reservation_status  
0           Check-Out  
1           Check-Out  
2           Check-Out  
3           Check-Out  
4           Check-Out  

[5 rows x 29 columns]

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [44]:
df.shape

(119390, 29)

Hasil `df.shape` menunjukkan bahwa DataFrame `df` memiliki 119,390 baris dan 29 kolom. Ini berarti DataFrame tersebut terdiri dari 119,390 data (baris) dan 29 fitur (kolom).

## Descriptive statistics

In [45]:
nums = []
cats = []

for column in df :
  if df[column].dtypes in ['int64', 'float64'] :
    nums.append(column)
  else :
    cats.append(column)

num_count = len(nums)
cat_count = len(cats)

print('Number of numerical columns:', num_count)
print('Number of categorical columns:', cat_count)
print('-' * 70)
print('Numerical columns:', nums)
print('Categorical columns:', cats)

Number of numerical columns: 20
Number of categorical columns: 9
----------------------------------------------------------------------
Numerical columns: ['is_canceled', 'lead_time', 'arrival_date_year', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_weekdays_nights', 'adults', 'children', 'babies', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'agent', 'company', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests']
Categorical columns: ['hotel', 'arrival_date_month', 'meal', 'city', 'market_segment', 'distribution_channel', 'deposit_type', 'customer_type', 'reservation_status']


Terdapat 20 kolom numerik ('nums') dan 9 kolom kategorikal ('cats') dalam DataFrame

In [46]:
df[nums].describe().T

count         mean         std      min  \
is_canceled                     119390.0     0.370416    0.482918     0.00   
lead_time                       119390.0   104.011416  106.863097     0.00   
arrival_date_year               119390.0  2018.156554    0.707476  2017.00   
arrival_date_week_number        119390.0    27.165173   13.605138     1.00   
arrival_date_day_of_month       119390.0    15.798241    8.780829     1.00   
stays_in_weekend_nights         119390.0     0.927599    0.998613     0.00   
stays_in_weekdays_nights        119390.0     2.500302    1.908286     0.00   
adults                          119390.0     1.856403    0.579261     0.00   
children                        119386.0     0.103890    0.398561     0.00   
babies                          119390.0     0.007949    0.097436     0.00   
is_repeated_guest               119390.0     0.031912    0.175767     0.00   
previous_cancellations          119390.0     0.087118    0.844336     0.00   
previous_bookings_not_canceled  119390.0     0.137097    1.497437     0.00   
booking_changes                 119390.0     0.221124    0.652306     0.00   
agent                           103050.0    86.693382  110.774548     1.00   
company                           6797.0   189.266735  131.655015     6.00   
days_in_waiting_list            119390.0     2.321149   17.594721     0.00   
adr                             119390.0   101.831122   50.535790    -6.38   
required_car_parking_spaces     119390.0     0.062518    0.245291     0.00   
total_of_special_requests       119390.0     0.571363    0.792798     0.00   

                                    25%       50%     75%     max  
is_canceled                        0.00     0.000     1.0     1.0  
lead_time                         18.00    69.000   160.0   737.0  
arrival_date_year               2018.00  2018.000  2019.0  2019.0  
arrival_date_week_number          16.00    28.000    38.0    53.0  
arrival_date_day_of_month          8.00    16.000    23.0    31.0  
stays_in_weekend_nights            0.00     1.000     2.0    19.0  
stays_in_weekdays_nights           1.00     2.000     3.0    50.0  
adults                             2.00     2.000     2.0    55.0  
children                           0.00     0.000     0.0    10.0  
babies                             0.00     0.000     0.0    10.0  
is_repeated_guest                  0.00     0.000     0.0     1.0  
previous_cancellations             0.00     0.000     0.0    26.0  
previous_bookings_not_canceled     0.00     0.000     0.0    72.0  
booking_changes                    0.00     0.000     0.0    21.0  
agent                              9.00    14.000   229.0   535.0  
company                           62.00   179.000   270.0   543.0  
days_in_waiting_list               0.00     0.000     0.0   391.0  
adr                               69.29    94.575   126.0  5400.0  
required_car_parking_spaces        0.00     0.000     0.0     8.0  
total_of_special_requests          0.00     0.000     1.0     5.0

Kesimpulan dari statistik deskriptif diatas sebagai berikut:

1. **is_canceled**: Rata-rata proporsi pembatalan reservasi adalah sekitar 37%, dengan nilai maksimum 100%.

2. **lead_time**: Rata-rata lead time (waktu antara pemesanan dan kedatangan) adalah sekitar 104 hari, dengan nilai maksimum 737 hari.

3. **arrival_date_year**: Data mencakup tahun-tahun 2017-2019, dengan mayoritas reservasi terjadi pada tahun 2018.

4. **arrival_date_week_number** dan **arrival_date_day_of_month**: Menunjukkan distribusi kedatangan tamu dalam minggu dan hari tertentu dalam bulan.

5. **stays_in_weekend_nights** dan **stays_in_weekdays_nights**: Rata-rata jumlah menginap di akhir pekan adalah sekitar 0.93 malam, sedangkan di hari kerja rata-rata sekitar 2.5 malam.

6. **adults**, **children**, **babies**: Rata-rata jumlah tamu dewasa adalah sekitar 1.86, dengan kebanyakan reservasi tidak melibatkan anak-anak atau bayi.

7. **previous_cancellations** dan **previous_bookings_not_canceled**: Mayoritas tamu tidak pernah melakukan pembatalan sebelumnya atau reservasi yang tidak dibatalkan.

8. **booking_changes**: Mayoritas reservasi tidak mengalami perubahan.

9. **agent** dan **company**: Terdapat data agen dan perusahaan yang melakukan reservasi, tetapi mayoritas data kosong (NaN).

10. **days_in_waiting_list**: Rata-rata hari dalam daftar tunggu adalah sekitar 2.32 hari, dengan nilai maksimum 391 hari.

11. **adr** (Average Daily Rate): Rata-rata tarif harian adalah sekitar 101.83.

12. **required_car_parking_spaces** dan **total_of_special_requests**: Mayoritas reservasi tidak membutuhkan tempat parkir tambahan atau memiliki permintaan khusus.

In [47]:
df[cats].describe()

hotel arrival_date_month       meal           city  \
count       119390             119390     119390         118902   
unique           2                 12          5            177   
top     City Hotel            October  Breakfast  Kota Denpasar   
freq         79330              13877      92310          48590   

       market_segment distribution_channel deposit_type customer_type  \
count          119390               119390       119390        119390   
unique              8                    5            3             4   
top         Online TA                TA/TO   No Deposit      Personal   
freq            56477                97870       104641         89613   

       reservation_status  
count              119390  
unique                  3  
top             Check-Out  
freq                75166

Kesimpulan dari statistik deskriptif diatas sebagai berikut:

1. **hotel**: Mayoritas data hotel adalah "City Hotel" (66.5%), sedangkan sisanya adalah "Resort Hotel".

2. **arrival_date_month**: Bulan dengan jumlah kedatangan tamu terbanyak adalah Oktober, sedangkan bulan dengan jumlah kedatangan terendah adalah Januari.

3. **meal**: Mayoritas reservasi termasuk makanan "Breakfast" (77.4%), sedangkan jenis makanan lainnya seperti "HB" (Half Board), "FB" (Full Board), dan "SC" (Self Catering) memiliki proporsi yang lebih kecil.

4. **city**: Kota dengan jumlah reservasi terbanyak adalah Kota Denpasar, yang mungkin merupakan lokasi hotel yang paling populer atau pusat destinasi wisata.

5. **market_segment**: Mayoritas reservasi berasal dari segment pasar "Online TA" (Travel Agents), diikuti oleh "Offline TA/TO" (Travel Agents/Tour Operators) dan "Groups".

6. **distribution_channel**: Distribusi pemesanan terbanyak dilakukan melalui "TA/TO" (Travel Agents/Tour Operators), diikuti oleh "Direct" (pemesanan langsung dengan hotel) dan "Corporate".

7. **deposit_type**: Sebagian besar reservasi tidak memerlukan deposit ("No Deposit"), sedangkan sebagian kecil memerlukan deposit "Non Refund".

8. **customer_type**: Mayoritas pelanggan adalah "Transient" (tamu yang tidak termasuk kelompok atau paket), diikuti oleh "Transient-Party" (tamu individu yang terkait dengan kelompok).

9. **reservation_status**: Mayoritas reservasi telah selesai ("Check-Out"), sedangkan sebagian kecil masih dalam proses ("Canceled" atau "No-Show").

## Data Preprocessing

In [48]:
pd.set_option('display.max_rows', 5000)
for i in df.columns:
    if i not in 'CustomerID':
        print(f'''Value count kolom {i}:''')
        print(df[i].value_counts())
        print()

Value count kolom hotel:
City Hotel      79330
Resort Hotel    40060
Name: hotel, dtype: int64

Value count kolom is_canceled:
0    75166
1    44224
Name: is_canceled, dtype: int64

Value count kolom lead_time:
0      6345
1      3460
2      2069
3      1816
4      1715
5      1565
6      1445
7      1331
8      1138
12     1079
11     1055
9       992
10      976
14      965
16      942
17      881
15      839
19      839
34      828
18      826
13      821
28      820
20      750
40      722
29      712
22      707
32      690
31      685
21      678
56      676
39      673
37      673
26      671
24      665
36      663
30      659
35      655
25      653
27      649
33      643
23      643
44      633
41      607
55      575
38      575
47      568
68      564
69      558
45      537
72      531
53      530
50      527
59      520
74      519
61      513
48      506
46      495
49      479
43      479
54      472
66      466
42      464
86      461
102     458
57      457
87      4

### Check Missing Value

In [51]:
missing_count = df.isnull().sum()

# Menghitung persentase missing value pada setiap kolom
missing_percentage = (df.isnull().sum() / len(df)) * 100

# Menggabungkan kedua Series ke dalam DataFrame
missing_df = pd.concat([missing_count, missing_percentage], axis=1)
missing_df.columns = ['Jumlah Missing Value', 'Persentase Missing (%)']

# Menampilkan kolom-kolom yang memiliki nilai null beserta persentase missing valuenya
missing_df[missing_df['Jumlah Missing Value'] > 0]

Jumlah Missing Value  Persentase Missing (%)
children                     4                0.003350
city                       488                0.408744
agent                    16340               13.686238
company                 112593               94.306893

In [52]:
white_space = []
for col in df.columns:
    for val in df[col]:
        if isinstance(val, str) and '  ' in val:
            white_space.append(val)

# Output
print(white_space)

[]


Kesimpulan dari check missing value:

1. Ditemukan 4 kolom dengan missing value: 'children' (4 missing values), 'city' (488 missing values), 'agent' (16,340 missing values), dan 'company' (112,593 missing values).

2. Persentase missing value tertinggi dimiliki oleh kolom 'company' dengan 94.31%, diikuti oleh 'agent' dengan 13.69%.

3. Tidak ditemukan nilai yang mengandung whitespace ganda dalam dataset.

## Handling Missing Value

Berdasarkan hasil check missing value, berikut adalah saran treatment yang tepat untuk null values:

1. **children**: Karena hanya terdapat 4 missing values pada kolom ini, kita bisa melakukan drop pada missing value.

2. **city**: Karena jumlah missing values cukup signifikan (488), dan mengingat kolom ini berisi informasi tentang kota, kita bisa mempertimbangkan untuk menggantinya dengan nilai yang paling sering muncul (mode).

3. **agent**: Dengan 16,340 missing values dan jumlah nilai yang berbeda sebanyak 333. Hal ini membuat menggantikan missing values dengan nilai rata-rata atau median tidaklah cocok, karena hal itu dapat mengaburkan informasi yang ada dalam kolom 'agent'. Sebagai gantinya, Anda dapat melakukan drop pada baris-baris yang memiliki nilai hilang tersebut agar tidak mengganggu analisis yang akan dilakukan pada kolom 'agent'.

4. **company**: Dengan 112,593 missing values, kemungkinan besar kolom ini tidak akan memberikan informasi yang signifikan. Sebaiknya, menghapus kolom ini dari dataset.

In [53]:
df.dropna(subset=['children'], inplace=True)
df['city'].fillna(df['city'].mode()[0], inplace=True)
df.dropna(subset=['agent'], inplace=True)
df.drop('company', axis=1, inplace=True)

In [54]:
df.isna().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_weekdays_nights          0
adults                            0
children                          0
babies                            0
meal                              0
city                              0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status          